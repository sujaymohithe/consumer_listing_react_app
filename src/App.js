import React from 'react';
import './App.css';
import ConsumerHome from './components/consumers/ConsumerHomeContainer';
import Header from './components/headers/Header';

function App() {
  return (
    <div className="App">
      <Header />
      <ConsumerHome />
    </div>
  );
}

export default App;
