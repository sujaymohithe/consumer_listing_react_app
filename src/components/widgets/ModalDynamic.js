import React from 'react';
import { Modal, Button } from 'react-bootstrap';

function ModalDynamic(props) {
  const { show, title, currentBudget, handleBudgetChange, budget_spent, onHide, saveBudget } = props;
  return (
    <Modal show={show} onHide={onHide}
      aria-labelledby="contained-modal-title-vcenter"
      centered
    >
      <Modal.Header closeButton>
        <Modal.Title id="contained-modal-title-vcenter">
          {title} 
        </Modal.Title>
      </Modal.Header>
      {/* Modal Body can be made dynamic by calling renderBody custom function based on modal title */}
      {/* Currently its not dynamic since we have only 1 modal popup in application */}
      <Modal.Body>
        <h5>Total Budget (€): </h5>
        <input id="total_budget" type="text" name="total_budget" value={currentBudget} onChange={handleBudgetChange} /><br/><br/>
        <h5>Budget Spent (€): {budget_spent && Number(budget_spent).toFixed(2)}</h5><br/>
        <p><b>Note - </b>Please enter value in German locale and upto 2 decimals only, for example -</p>
        <ul>
          <li>10.000,00</li>
          <li>4.500,60</li>
          <li>451,38</li>
        </ul>
      </Modal.Body>
      <Modal.Footer>
        <Button variant="secondary" onClick={onHide} id="closeBtn">Close</Button>
        <Button variant="primary" onClick={saveBudget}>Save Changes</Button>
      </Modal.Footer>
    </Modal >
  );
}

export default ModalDynamic;