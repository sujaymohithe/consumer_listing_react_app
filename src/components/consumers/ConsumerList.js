import React from 'react';
import ModalDynamic from '../widgets/ModalDynamic';
import toastr from "toastr";

//Created below class component, can be converted to functional component if required. Also can be used Reacts Hooks.
class ConsumerList extends React.Component {

    constructor(props) {
        super(props);
        this.state = {
            show: false,
            currentBudget: 0,
            selectedRowData: []
        }
    }

    //As dataset is not in German locale, using below ECMAScript Internationalization API to convert budget values into German locale
    getGermanFormat(num) {
        return Intl.NumberFormat('de-DE', { style: 'currency', currency: 'EUR' }).format(num)
    }

    //below function is used to show all Martian consumers in a table
    populateMartianConsumerTable(consumer, index) {
        return (
            <tr key={index} onClick={(e) => this.handleShow(consumer)}>
                <td>{consumer.name}</td>
                <td>{consumer.date_of_first_purchase}</td>
                <td>{this.getGermanFormat(consumer.budget)}</td>
                <td>{this.getGermanFormat(consumer.budget_spent)}</td>
                <td>{this.getGermanFormat(consumer.budget - consumer.budget_spent)}</td>
            </tr>
        );
    }

    //to hide pop-up
    handleClose() {
        this.setState({ show: false })
    }

    //to show pop up and to prepopulate total budget in pop up by pushing into state
    handleShow(consumer) {
        let currentBudgetGermantFormat = this.getGermanFormat(consumer.budget).replace('â‚¬', '').trim();
        this.setState({ show: true, currentBudget: currentBudgetGermantFormat, selectedRowData: consumer })
    }

    //to handle input for total budget textbox in pop up
    handleBudgetChange(e) {
        this.setState({ currentBudget: e.target.value });
    }

    //to check total budget to be greater than or equal to budget spent
    isTotalBudgetGreaterThanBudgetSpent(changedBudgetValue, budgetSpentValue) {
        let selectedConsumerSpentBudget = Number(budgetSpentValue.toFixed(2));
        return changedBudgetValue >= selectedConsumerSpentBudget;
    }

    validateInputForGermanFormat(changedBudgetValue) {
        //to check entered budget value meets the German locale standard
        var regex = /^[1-9]\d*(((.\d{3}){0,})?(\,\d{0,2})?)$/;
        return regex.test(changedBudgetValue);
    }

    //to save total budget input change into dataset
    saveBudget() {
        let selectedRow = this.state.selectedRowData, ref = this;
        let changedBudgetValue = this.state.currentBudget;
        if (ref.validateInputForGermanFormat(changedBudgetValue)) {
            //In JS, it is recommended to handle currency as normal integer for arithmatic opearions. Hence converting german format to integar format
            let changedBudgetIntValue = changedBudgetValue.replace(/\./g, '').replace(',', '.');
            if (ref.isTotalBudgetGreaterThanBudgetSpent(changedBudgetIntValue, selectedRow.budget_spent)) {
                ref.props.saveConsumerBudget(selectedRow, changedBudgetIntValue)
                    .then((result) => {
                        if (result === 200) {
                            this.props.getConsumerList();
                            toastr.success("Saved successfully.");
                            ref.setState({ show: false });
                        }
                        else {
                            toastr.error("Error has occured !");
                        }
                    })
                    .catch(error => {
                        toastr.error('Error occured, contact admin.');
                    });;
            }
            else {
                toastr.warning('The value must not be less than the budget spent.');
            }
        }
        else {
            toastr.warning('Refer note section, please enter budget in correct german locale and upto 2 decimal points. ');
        }
    }

    render() {
        return (
            <div>
                <br />
                <h2>All Martian Consumers</h2>
                <table className="table table-striped">
                    <thead>
                        <tr>
                            <th>Company</th>
                            <th>Date of first purchase</th>
                            <th>Total budget</th>
                            <th>Budget spent</th>
                            <th>Budget left</th>
                        </tr>
                    </thead>
                    <tbody>
                        {this.props.consumers && this.props.consumers.map(this.populateMartianConsumerTable, this)}
                    </tbody>
                </table>

                <ModalDynamic show={this.state.show}
                    onHide={(e) => this.handleClose()} title={this.state.selectedRowData.name} budget_spent={this.state.selectedRowData.budget_spent} currentBudget={this.state.currentBudget}
                    handleBudgetChange={(e) => this.handleBudgetChange(e)}
                    saveBudget={() => this.saveBudget()} />
            </div>
        );
    }
}

export default ConsumerList