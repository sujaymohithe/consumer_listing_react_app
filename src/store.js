// Imports
import { compose, combineReducers } from 'redux'
import { createStore, applyMiddleware } from 'redux'
import thunk from 'redux-thunk'

// App Imports
import * as consumersData from './reducers/consumers/ConsumersReducer';

// App Reducer - currently there is 1 reducer, if application grows complex then we can have multiple reducers and later
// combine all using below method
const appReducer = combineReducers({
    ...consumersData
})

// Root Reducer
export const rootReducer = (state, action) => {
    if (action.type === 'RESET') {
        state = undefined
    }
    return appReducer(state, action)
}

// Load initial state from server side
let initialState
if(typeof window !== 'undefined') {
    initialState = window.__INITIAL_STATE__
    delete window.__INITIAL_STATE__
}

// Store
export const store = createStore(
    rootReducer,
    initialState,

    compose(
        applyMiddleware(thunk),
    )
)
