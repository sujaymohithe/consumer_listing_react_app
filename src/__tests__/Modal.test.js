import React from 'react';
import { shallow, mount } from 'enzyme';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { Provider } from "react-redux";
import App from '../App';
import { store } from '../store';
import ConsumerHome from '../components/consumers/ConsumerHomeContainer';
import ConsumerList from '../components/consumers/ConsumerList';
import ModalDynamic from '../components/widgets/ModalDynamic';
import { Modal, Button } from 'react-bootstrap';

Enzyme.configure({ adapter: new Adapter() });

describe('<ModalDynamic/>', () => {
    it('renders without parent component', () => {
        const props = {
            show: true
        };
        const ModalWrapperComponent = shallow(<ModalDynamic {...props} />);
        expect(ModalWrapperComponent).toMatchSnapshot();
    });

    it('renders the modal component successfully', () => {
        const wrapper = mount(
            <Provider store={store}>
                <App />
            </Provider>);
        expect(wrapper.find(ConsumerHome).find(ConsumerList).find(ModalDynamic).length).toEqual(1);
    })

    it('should have input field', () => {
        const props = {
            show: true
        };
        const modalWrapper = shallow(<ModalDynamic {...props} />);
        expect(modalWrapper.find('input').length).toEqual(1);
    })

    it('should check format of number', () => {
        const props = {
            show:true
        },
        modalWrapper = shallow(<ConsumerList />);
        expect(modalWrapper.instance().getGermanFormat(451.3754)).toEqual('€451.38');
    })

    it('should check wrong format of number', () => {
        const modalWrapper = shallow(<ConsumerList />);
        expect(modalWrapper.instance().getGermanFormat('test123')).toEqual('€NaN');
    })

    it('should check wrong format of number', () => {
        const wrapper = shallow(<ConsumerList />);
        expect(wrapper.instance().validateInputForGermanFormat('111,00')).toEqual(true);
    })

    it('should check input format of modal text entry to be in German format', () => {
        const wrapper = shallow(<ConsumerList />);
        expect(wrapper.instance().validateInputForGermanFormat('111,00')).toEqual(true);
    })

    it('should check input format of modal text entry to be in German format', () => {
        const wrapper = shallow(<ConsumerList />);
        expect(wrapper.instance().validateInputForGermanFormat('111.111.111,00')).toEqual(true);
    })

    it('should check input format of modal text entry to be in German format', () => {
        const wrapper = shallow(<ConsumerList />);
        expect(wrapper.instance().validateInputForGermanFormat('111.11.111,00')).toEqual(false);
    })

    it('check show prop value and type', () => {
        const props = {
            show: true
        },
            ModalWrapperComponent = shallow(<ModalDynamic {...props} />);
        expect(ModalWrapperComponent.props().show).toEqual(true);
    });
    
    it('render correct onHide prop type', () => {
        const props = {
            onHide: () => { }
        },
            ModalWrapperComponent = shallow(<ModalDynamic {...props} />);
        expect(ModalWrapperComponent.props().onHide).toBeInstanceOf(Function)
    });

    it('render correct title', () => {
        const props = {
            title: 'Martian Firma'
        };
        let ModalWrapperComponent = shallow(<ModalDynamic {...props} />).find('#contained-modal-title-vcenter');
        expect(ModalWrapperComponent.text()).toEqual('Martian Firma');
    });

    it('render correct total budget', () => {
        const props = {
            currentBudget: '10.000,00',
            budget_spent: '4.500,60'
        };
        let ModalWrapperComponent = shallow(<ModalDynamic {...props} />);
        const input = ModalWrapperComponent.find('#total_budget');
        expect(input.props().value).toEqual("10.000,00");
    });

    it('opens the modal upon row click', () => {
        const props = {
            consumers: [{ "id": 1, "name": "Martian Firma" }, { "id": 2, "name": "Solar Firma" }]
        };
        const wrapper = mount(
            <Provider store={store}>
                <ConsumerList {...props} />
            </Provider>);
        const tr = wrapper.find('table tbody tr').first();
        tr.simulate('click');
        //check whether modal opens up
        expect(wrapper.find(Modal).prop('show')).toEqual(true);
    })

    it('Test button click event', () => {
        const mockCallBack = jest.fn();
        const button = shallow((<Button onClick={mockCallBack}>Ok!</Button>));
        button.find('button').simulate('click');
        expect(mockCallBack.mock.calls.length).toEqual(1);
    })

    it('closes the modal upon close button click', () => {
        const props = {
            consumers: [{ "id": 1, "name": "Martian Firma" }, { "id": 2, "name": "Solar Firma" }]
        };
        const wrapper = mount(
            <Provider store={store}>
                <ConsumerList {...props} />
            </Provider>);
        const tr = wrapper.find('table tbody tr').first();
        tr.simulate('click');
        //check whether modal closes up using property show
        const closebutton = wrapper.find('#closeBtn').at(0);
        closebutton.simulate('click');
        expect(wrapper.find(Modal).prop('show')).toEqual(false);
    })

})