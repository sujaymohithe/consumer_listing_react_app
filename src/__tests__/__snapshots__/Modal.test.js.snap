// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ModalDynamic/> renders without parent component 1`] = `ShallowWrapper {}`;
